solidity/random-beacon/contracts/RandomBeacon.sol

```diff
@@ -548,6 +589,39 @@ contract RandomBeacon is Ownable {
         }
     }
 
+    /// @notice Punishes the given operators by kicking them out from the
+    ///         sortition pool and blocking their re-join for a given period
+    ///         of time.
+    /// @dev By the way, this function releases gas deposits made by operators
+    ///      during their registration. See `registerOperator` function. This
+    ///      action makes punishments cheaper gas-wise.
+    /// @param operators Addresses of punished operators.
+    /// @param punishmentDuration Duration of the punishment period in seconds.
+    function punishOperators(
+        address[] memory operators,
+        uint256 punishmentDuration
+    ) internal {
+        for (uint256 i = 0; i < operators.length; i++) {
+            address operator = operators[i];
+
+            // Set the punishment regardless the operator is actually registered
+            // in the sortition pool.
+            /* solhint-disable not-rely-on-time */
+            // slither-disable-next-line reentrancy-benign
+            punishedOperators[operator] = block.timestamp + punishmentDuration;
+            /* solhint-enable not-rely-on-time */
+
+            // Perform operations on sortition pool only in case the operator
+            // is registered. Otherwise, the punishment set above will prevent
+            // the operator to join the pool in future.
+            if (sortitionPool.isOperatorInPool(operator)) {
+                gasStation.releaseGas(operator);
+                // TODO: Is it possible to kick out operator when pool is locked?
+                sortitionPool.removeOperator(operator);
```

Comment:
Yeah, after spending some time in the sortition pool code I think we can theoretically block the entry submission if we try to kick out an operator who has already left the sortition pool. Your proposition sounds good. The only thing to consider is the fact that we plan to use `sortitionPool.removeOperators` to remove multiple operators in one call. In that case, if we surround that call with try-catch and we fail during the removal of one operator, the rest of operators avoid the punishment in that case. Maybe we should handle such a case at the sortition pool level? Curious about your thoughts @dimpar and @pdyraga.

Reply:
Or we can try to capture the "id" of that failed operator and retry starting from the next id? Although it might be too much and moving that check to the sortition pool might be a solution.

Reply:
I think the recent changes introduced in `punishOperators` method solve the problem we discuss here. Only actual pool members are passed to `sortitionPool.removeOperators` so there is no risk we would try to remove an operator who already left the pool and we'll revert in the result. Just keep in mind the current implementation of `punishOperators` can be suboptimal and we will be able (probably) to make it cheaper in the near future once all pending changes developed in the sortition pool codebase are ready.